### PR TITLE
Store and load comments in Datastore and fix implementation of Optional in getParameter()

### PR DESCRIPTION
**Pull request by officialcwang**
Finish the fifth tutorial for week 3.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -26,28 +32,46 @@
...
-    String text = getParameter(request, "text-input", "");
+    Optional<String> textOptional = getParameter(request, "text-input");
+    String text = textOptional.orElse("");
 
     // Respond with the result.
     String json = gson.toJson(text);
     comments.add(json);
 
+    // Store the comment.
+    Entity commentEntity = new Entity("Comment");
```
Comment (chen-dawn) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:70]: Prefer to make some of these strings constants
`private static final String COMMENT = "Comment"` etc

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -58,11 +82,8 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
    *         was not specified by the client
    */
-  Optional String getParameter(HttpServletRequest request, String name, String defaultValue) {
-    String value = request.getParameter(name);
-    if (value == null) {
-      return defaultValue;
-    }
-    return value;
+  private Optional<String> getParameter(HttpServletRequest request, String name) {
+    Optional<String> result = Optional.ofNullable(request.getParameter(name));
```
Comment (chen-dawn) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:86]: You can just return the output directly? Don't need the `result` variable.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -26,28 +32,46 @@
...
   /**
-   * Retrieve inputted comments and redirect the result to the page.
+   * Retrieve and store inputted comments and redirect the result to the page.
    */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Get the input from the form.
-    String text = getParameter(request, "text-input", "");
+    Optional<String> textOptional = getParameter(request, "text-input");
+    String text = textOptional.orElse("");
```
Comment (chen-dawn) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:63]: Maybe not add the comment if it's not present? Perhaps add some validation on the front end (e.g. disable comment button when there's no text entered?)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -26,28 +32,46 @@
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
```
Comment (chen-dawn) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:34]: Maybe add some tests? (though we can wait till week 5 which is about testing).

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -46,6 +47,16 @@ function createListElement(text) {
   return liElement;
 }
 
+/** Checks if the user has inputted any text upon clicking submit. */
+function isInputEmpty() {
+  const comment = document.getElementById('text-input').value;
+  if (comment == null || comment === '') {
```
Comment (luli0521) [inline portfolio/src/main/webapp/script.js:53]: if (!comment || comment === '')
therefore you check whether comment is null or undefined.

And do you want to check whether the comment contains only spaces?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -26,28 +32,48 @@
...
-  private List<String> comments = new ArrayList<>();
+  private List<String> comments;
   private Gson gson = new Gson();
+  private static final String COMMENT = "Comment";
+  private static final String TEXT = "text";
```
Comment (luli0521) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:38]: These names are not good enough. The fields don't represent comment or text content, but the column/field names when you interact with DataStore. The followings are better names:

COMMENT_COLUMN
COMMENT_FIELD
COMMENT_COLUMN_NAME
COMMENT_FIELD_NAME

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -46,6 +47,26 @@ function createListElement(text) {
...
+  const comment = document.getElementById('text-input').value;
+  if (!comment || comment === '' || isInputOnlySpaces(comment)) {
+    alert('Please enter a valid comment.');
+    return false;
+  }
+  return true;
+}
+
+/** Checks if the user has inputted only spaces. */
+function isInputOnlySpaces(input) {
```
Comment (chen-dawn) [inline portfolio/src/main/webapp/script.js:61]: prefer to use a regex pattern, then you can check for spaces, new lines, tabs... (or if it contains at least one alphaneumeric character)

**Pull request merged**